### PR TITLE
metric: add ttl expire metric

### PR DESCRIPTION
**Pull request by Smityz**
Signed-off-by: Smityz <smityz@qq.com>

### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/15873

What's Changed:

```commit-message
metric: add ttl expire metric
```
image

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```

### CODE REVIEW

File: src/server/ttl/ttl_compaction_filter.rs [hunk 1]
```diff
@@ -11,9 +11,23 @@ use engine_rocks::{
     RocksTtlProperties,
 };
 use engine_traits::raw_ttl::ttl_current_ts;
+use prometheus::*;
 
 use crate::server::metrics::TTL_CHECKER_ACTIONS_COUNTER_VEC;
 
+lazy_static! {
+    pub static ref TTL_EXPIRE_KV_SIZE_COUNTER: IntCounter = register_int_counter!(
+        "tikv_ttl_expire_kv_size_total",
```
Comment (afeinberg) [inline src/server/ttl/ttl_compaction_filter.rs:20]: nit: would it make sense to use ``raw`` in the string name? e.g., "``ikv_ttl_raw_expire_kv_{size,entry}_total``

File: src/server/ttl/ttl_compaction_filter.rs [hunk 2]
```diff
@@ -83,7 +97,11 @@ impl<F: KvFormat> CompactionFilter for TtlCompactionFilter<F> {
             Ok(RawValue {
                 expire_ts: Some(expire_ts),
                 ..
-            }) if expire_ts <= self.ts => CompactionFilterDecision::Remove,
+            }) if expire_ts <= self.ts => {
+                TTL_EXPIRE_KV_SIZE_COUNTER.inc_by(key.len() as u64 + value.len() as u64);
```
Comment (pingyu) [inline src/server/ttl/ttl_compaction_filter.rs:101]: Suggest to accumulate the metrics in member variables first, then add to the counters on dropping `TtlCompactionFilter`. That would slightly improve performance as prometheus counters are atomic variables unerlying.

File: src/server/ttl/ttl_compaction_filter.rs (hunk 1 above)
Comment (Smityz) [reply src/server/ttl/ttl_compaction_filter.rs:20]: The other metrics in `TTL` navigation are named as tikv_ttl_*, like https://github.com/tikv/tikv/blob/master/src/server/metrics.rs#L365, maybe we should keep them consistent

File: src/server/ttl/ttl_compaction_filter.rs [hunk 3]
```diff
@@ -56,6 +67,28 @@ impl<F: KvFormat> CompactionFilterFactory for TtlCompactionFilterFactory<F> {
...
     ts: u64,
     _phantom: PhantomData<F>,
+    expire_entry: u64,
+    expire_size: u64,
+}
+
+impl<F: KvFormat> Drop for TtlCompactionFilter<F> {
+    fn drop(&mut self) {
+        // Accumulate counters would slightly improve performance as prometheus counters
+        // are atomic variables unerlying
```
Comment (pingyu) [inline src/server/ttl/ttl_compaction_filter.rs:77]:
```suggestion
        // are atomic variables underlying
```

**Review by pingyu — APPROVED** [review]
LGTM~

File: src/server/ttl/ttl_compaction_filter.rs [hunk 4]
```diff
@@ -83,7 +116,11 @@ impl<F: KvFormat> CompactionFilter for TtlCompactionFilter<F> {
             Ok(RawValue {
                 expire_ts: Some(expire_ts),
                 ..
-            }) if expire_ts <= self.ts => CompactionFilterDecision::Remove,
+            }) if expire_ts <= self.ts => {
+                self.expire_size += key.len() as u64 + value.len() as u64;
+                self.expire_entry += 1;
```
Comment (bufferflies) [inline src/server/ttl/ttl_compaction_filter.rs:121]: how about rename `expire_count`?

**Review by bufferflies — APPROVED** [review]
rest lgtm

File: src/server/ttl/ttl_compaction_filter.rs (hunk 4 above)
Comment (Smityz) [reply src/server/ttl/ttl_compaction_filter.rs]: done

**Comment by pingyu** [comment]
@Smityz Thanks for your contribution !

**Comment by Smityz** [comment]
> @Smityz Thanks for your contribution !

To be continued!

**Pull request merged**